pineappl_py/pineappl/grid.py

```diff
@@ -268,7 +274,7 @@ def convolute_eko(self, operators, lumi_id_types="pdg_mc_ids", order_mask=()):
         alphas_values = [op["alphas"] for op in operators["Q2grid"].values()]
         return FkTable(
             self.raw.convolute_eko(
-                np.array(operators["q2_ref"]),
+                operators["q2_ref"],
                 np.array(alphas_values),
```

Comment:
I wonder where to cast:
- for `targetpids` you're doing it here below, but
- for `alphas_values` you're not doing it
- however it will eventually happen as Rust is requiring it here https://github.com/N3PDF/pineappl/blob/6806ad0f74ef31ee8cfdc27fbc96dea5f6dded66/pineappl_py/src/grid.rs#L330 (of course due to https://github.com/N3PDF/pineappl/blob/6806ad0f74ef31ee8cfdc27fbc96dea5f6dded66/pineappl/src/grid.rs#L371)

this is only one example out of many

Reply:
> You're also mixing two different functions, `convolute_with_one` and `convolute_eko`, so I'm pretty confused...

Mmm, I think, I pointed to  `convolute_eko` always - in case I didn't sorry :upside_down_face: - as said the thing is general

> I didn't understand what do you mean by _where to cast_.

I mean the difference between this line:
https://github.com/N3PDF/pineappl/blob/6806ad0f74ef31ee8cfdc27fbc96dea5f6dded66/pineappl_py/pineappl/grid.py#L279
- with an explicit cast to int
- so numpy will cast here
- and e.g. `[1.,2.]` will result into `np.array([1,2])`

and this line:
https://github.com/N3PDF/pineappl/blob/6806ad0f74ef31ee8cfdc27fbc96dea5f6dded66/pineappl_py/pineappl/grid.py#L278
- with *no* explict cast here
- so Rust will cast upon consumption (I guess - since I don't know how PyO3 is working)
- as it is requested here https://github.com/N3PDF/pineappl/blob/6806ad0f74ef31ee8cfdc27fbc96dea5f6dded66/pineappl_py/src/grid.rs#L330

The concern is not about passing lists or array, but about the list content type - in Rust language: `PyReadonlyArray1<f64>`

Explicit suggestion: we may change the line to `np.array(alphas_values, dtype=np.float)`

Reply:
I only have to cast to integers, because they are float by default

Reply:
I don't believe so: https://numpy.org/doc/stable/reference/generated/numpy.array.html?highlight=array#numpy.array
"The desired data-type for the array. If not given, then the type will be determined as the minimum type required to hold the objects in the sequence."

Reply:
That is: you need only one float to make it float

- alphas is a positive number smaller than one
- same goes for x

Reply:
P.S.: @felixhekhorn the moment you have a reasonable example breaking, I'll be happy to cast immediately for that one :)

Reply:
I don't - I'd just thought about following the general rule: "I'd rather cast myself then at a random place" ... fine, I accept